server/src/main/java/io/pravega/schemaregistry/service/SchemaRegistryService.java

```diff
@@ -0,0 +1,924 @@
+        log.info("getGroupProperties called for group {}.", group);
+        return store.getGroupProperties(namespace, group)
+                    .whenComplete((r, e) -> {
+                        if (e == null) {
+                            log.info("Group {} properties.", group);
+                        } else {
+                            log.warn("getGroupProperties for group {} request failed with error", group, e);
+                        }
+                    });
+
+    }
+
+    /**
+     * Update group's schema compatibility policy. If previous rules are sent, a conditional update is performed.
+     *
+     * @param namespace       namespace for which the request is scoped to.
+     * @param group           Name of group.
+     * @param compatibility New compatibility for the group.
+     * @param previousCompatibility   Previous rules compatibility for the group. If null, unconditional update is performed.
+     * @return CompletableFuture which is completed when compatibility policy update completes.
+     */
+    public CompletableFuture<Void> updateCompatibility(String namespace, String group, Compatibility compatibility,
+                                                       @Nullable Compatibility previousCompatibility) {
+        Preconditions.checkArgument(group != null);
+        Preconditions.checkArgument(compatibility != null);
+        log.info("updateCompatibility called for group {}. New compatibility {}", group, compatibility);
+        return RETRY.runAsync(() -> store.getGroupEtag(namespace, group)
+                                         .thenCompose(pos -> {
+                                             return store.getGroupProperties(namespace, group)
+                                                         .thenCompose(prop -> {
+                                                             if (previousCompatibility == null) {
+                                                                 return store.updateCompatibility(namespace, group, pos, compatibility);
+                                                             } else {
+                                                                 if (previousCompatibility.equals(prop.getCompatibility())) {
+                                                                     return store.updateCompatibility(namespace, group, pos, compatibility);
+                                                                 } else {
+                                                                     throw new PreconditionFailedException("Conditional update failed");
+                                                                 }
+                                                             }
+                                                         });
+                                         })
+                                         .whenComplete((r, e) -> {
+                                             if (e == null) {
```

Comment:
On second thought, can logging be done without running this in a separate completion stage?

Reply:
Not really.